Signal/src/call/SignalCall.swift

```diff
@@ -39,51 +42,97 @@ protocol CallObserver: class {
 
     // Distinguishes between calls locally, e.g. in CallKit
     let localId: UUID
-    
+
     var hasVideo = false {
         didSet {
-            Logger.debug("\(TAG) hasVideo changed: \(oldValue) -> \(hasVideo)")
-            for observer in observers {
-                observer.value?.hasVideoDidChange(call: self, hasVideo: hasVideo)
+            // This should only occur on the signaling queue.
+            objc_sync_enter(self)
```

Comment:
> I've protected all of its state through synchronization.

I'm not sure I understand that claim. Accessing it from the main thread while it's mutated on the signalingQueue would still give an inconsistent state. In other words - we're not doing any locking on access. I think I understand your concern there, but I don't see how this addresses it.

My question, reworded: If all the methods you've added `objc_sync` to are already being run on the same serial queue, then what difference could adding `objc_sync` possibly make?

> More generally, I'd like to propose that we have an explicit understanding that all state should only be accessed from one thread.

Sounds like a good rule of thumb, but I'm not sure I'm comfortable committing to this 100%. For example, I'm not sure that meshes with YapDatabase usage conventions, but maybe I'm misunderstanding. In any case, I appreciate your insight, and look forward to your highlighting specific related examples in code reviews going forward.